ldclient/client.py

```diff
@@ -0,0 +1,228 @@
+
+    def upsert(self, key, feature):
+        try:
+            self._lock.lock()
+            f = self._features.get(key)
+            if f is None or f['version'] < feature['version']:
+                self._features[key] = f
+        finally:
+            self._lock.unlock()
+
+    @property
+    def initialized(self):
+        try:
+            self._lock.rlock()
+            return self._initialized
+        finally:
+            self._lock.runlock()
+
+
+class LDClient(object):
+    def __init__(self, api_key, config=None):
+        check_uwsgi()
+        self._api_key = api_key
+        self._config = config or Config.default()
+        self._session = CacheControl(requests.Session())
+        self._queue = queue.Queue(self._config.capacity)
+        self._consumer = None
+        self._offline = False
+        self._lock = Lock()
+
+        self._store = config.feature_store_class()
+        """ :type: FeatureStore """
+
+        self._feature_requester = config.feature_requester_class(api_key, config)
+        """ :type: FeatureRequester """
+
+        self._stream_processor = None
+        if self._config.stream:
+            self._stream_processor = config.stream_processor_class(api_key, config, self._store)
+            self._stream_processor.start()
+
+    @property
+    def api_key(self):
+        return self._api_key
+
+    def _check_consumer(self):
+        with self._lock:
+            if not self._consumer or not self._consumer.is_alive():
+                self._consumer = self._config.consumer_class(self._queue, self._api_key, self._config)
+                self._consumer.start()
+
+    def _stop_consumers(self):
+        if self._consumer and self._consumer.is_alive():
+            self._consumer.stop()
+        if self._stream_processor and self._stream_processor.is_alive():
+            self._stream_processor.stop()
+
+    def _send(self, event):
+        if self._offline:
+            return
+        self._check_consumer()
+        event['creationDate'] = int(time.time() * 1000)
+        if self._queue.full():
+            log.warning("Event queue is full-- dropped an event")
+        else:
+            self._queue.put(event)
+
+    def track(self, event_name, user, data=None):
+        self._send({'kind': 'custom', 'key': event_name, 'user': user, 'data': data})
+
+    def identify(self, user):
+        self._send({'kind': 'identify', 'key': user['key'], 'user': user})
+
+    def set_offline(self):
+        self._offline = True
+        self._stop_consumers()
+
+    def set_online(self):
+        self._offline = False
+        self._check_consumer()
+
+    def is_offline(self):
+        return self._offline
+
+    def flush(self):
+        if self._offline:
+            return
+        self._check_consumer()
+        return self._consumer.flush()
+
+    def get_flag(self, key, user, default=False):
+        return self.toggle(key, user, default)
+
+    def toggle(self, key, user, default=False):
```

Comment:
This function is the one significant change here.  The implementation is simpler in that it isolates the actual retrieval from the rest, but more complicated in the callback function.  That func is necessary to support returning defers/tasks for async and the actual value for normal usage.